docs/source/tutorial/dags.rst

```diff
@@ -0,0 +1,287 @@
+quantum circuit so it can be executed on real hardware. For example, a
+simple transformation pass is to translate the quantum gates in the circuit
+to the set of gates allowed on a device. In the above example if we were to
+attempt to run that on a QPU that didn't natively support the ``H`` quantum
+gate we'd have to translate that to an equivalent series of instructions
+that the hardware actually supported. A simplified view of how this is
+performed is:
+
+.. jupyter-execute::
+
+    # Equivalency matrix
+    translation_matrix = {"h": ["rz(pi/2)", "sx", "rz(pi/2)"]}
+    # Insructions natively supported on target QPU
+    hardware_instructions = {"measure", "cx", "sx", "rz", "x"}
+
+    # Iterate over instructions in order and replace gates outside of native
+    # instruction set with a subcircuit from the translation matrix
+    for gate_index in rx.topological_sort(dag):
+        if gate_index not in in_nodes and gate_index not in out_nodes:
+            if dag[gate_index] not in hardware_instructions:
+                edge_val = dag.out_edges(gate_index)[0][2]
+                equivalent_subcircuit = rx.PyDiGraph()
+                count = 0
+                for node in translation_matrix[dag[gate_index]]:
+                    if count == 0:
+                        equivalent_subcircuit.add_node(node)
+                    else:
+                        equivalent_subcircuit.add_child(count - 1, node, edge_val)
+                    count += 1
+
+                def map_fn(source, target, weight):
+                    if source == gate_index:
+                        return len(equivalent_subcircuit) - 1
+                    else:
+                        return 0
+
+                dag.substitute_node_with_subgraph(
+                    gate_index,
+                    equivalent_subcircuit,
+                    map_fn
+                )
+
+    graphviz_draw(
+        dag,
+        node_attr_fn=lambda node: {"label": str(node)},
+        edge_attr_fn=lambda edge: {"label": str(edge)}
+    )
+
+Another example of how the compiler in Qiskit operates on a DAG is to perform
+analysis to find all the instances of single qubit gates that are executed in
+series. This series of quantum gates can be analyzed and often simplified
+into a shorter sequence of gates. A simplified example of this analysis is:
+
+.. jupyter-execute::
+
+    bit_nodes = {"q_0", "q_1", "c_0", "c_1"}
+
+    def filter_fn(node):
+        # Don't collect input or output nodes
+        if node in bit_nodes:
+            return False
+        # Don't include 2 qubit gates
+        if node == "cx":
+            return False
+        # Ignore non-unitary operations
+        if node == "measure":
+            return False
+        return True
+
+    runs = rx.collect_runs(dag, filter_fn)
+    print(runs)
+
+With this we have the DAG nodes that make up a series of 1 qubit gates that
+we can analyze and attempt to simplify. Skipping the details of the
```

Comment:
I like the name drop for `contract_nodes` (since I wrote it :D), but I wonder if this just gets too complicated for the purpose of this tutorial. Perhaps it makes sense to stop at finding the series of 1-qubit gates to simplify (i.e. stop the tutorial here without showing how to actually do the simplification).

Reply:
I'll defer to you here, I thought it was good to show the full path of finding runs in the circuit and then showing the simplification because I thought it might be confusing to say we can find these things in the DAG easily but not show what we do with it afterwards. But I don't feel strongly about it and if you think this is too much detail we can drop it.

Reply:
Makes sense. I suggested a change just now that I think would make this a little easier to read, while keeping the content.